File: app/src/main/java/de/westnordost/streetcomplete/data/osmnotes/CreateNoteDao.java
package de.westnordost.streetcomplete.data.osmnotes;

import android.content.ContentValues;
import android.database.Cursor;
import android.database.sqlite.SQLiteDatabase;
import android.database.sqlite.SQLiteOpenHelper;

import java.util.ArrayList;
import java.util.List;

import javax.inject.Inject;

import de.westnordost.streetcomplete.data.WhereSelectionBuilder;
import de.westnordost.osmapi.map.data.BoundingBox;
import de.westnordost.osmapi.map.data.Element;
import de.westnordost.osmapi.map.data.OsmLatLon;
import de.westnordost.streetcomplete.util.Serializer;

public class CreateNoteDao
{
	protected final SQLiteOpenHelper dbHelper;
	private final Serializer serializer;

	@Inject public CreateNoteDao(SQLiteOpenHelper dbHelper, Serializer serializer)
	{
		this.dbHelper = dbHelper;
		this.serializer = serializer;
	}

	public boolean add(CreateNote note)
	{
		SQLiteDatabase db = dbHelper.getWritableDatabase();
		ContentValues values = new ContentValues();
		values.put(CreateNoteTable.Columns.LATITUDE, note.position.getLatitude());
		values.put(CreateNoteTable.Columns.LONGITUDE, note.position.getLongitude());
		if(note.elementType != null)
		{
			values.put(CreateNoteTable.Columns.ELEMENT_TYPE, note.elementType.name());
		}
		if(note.elementId != null)
		{
			values.put(CreateNoteTable.Columns.ELEMENT_ID, note.elementId);
		}
		if (note.imagePaths != null)
		{
			values.put(CreateNoteTable.Columns.IMAGE_PATHS, serializer.toBytes(note.imagePaths));
		}
		values.put(CreateNoteTable.Columns.TEXT, note.text);
		if (note.questTitle != null)
		{
			values.put(CreateNoteTable.Columns.QUEST_TITLE, note.questTitle);
		}

		long rowId = db.insert(CreateNoteTable.NAME, null, values);

		if(rowId != -1)
		{
			note.id = rowId;
			return true;
		}
		return false;
	}

	public CreateNote get(long id)
	{
		SQLiteDatabase db = dbHelper.getWritableDatabase();
		String where = CreateNoteTable.Columns.ID + " = " + id;

		try (Cursor cursor = db.query(CreateNoteTable.NAME, null, where, null, null, null, null, "1"))
		{
			if (!cursor.moveToFirst()) return null;
			return createObjectFrom(cursor);
		}
	}


	public int getCount()
	{
		SQLiteDatabase db = dbHelper.getReadableDatabase();

		try (Cursor cursor = db.query(CreateNoteTable.NAME, new String[]{"COUNT(*)"},
				null, null, null, null, null, null))
		{
			cursor.moveToFirst();
			return cursor.getInt(0);
		}
	}

	public boolean delete(long id)
	{
		SQLiteDatabase db = dbHelper.getWritableDatabase();
		return db.delete(CreateNoteTable.NAME, CreateNoteTable.Columns.ID + " = " + id, null) == 1;
	}

	public List<CreateNote> getAll(BoundingBox bbox)
	{
		SQLiteDatabase db = dbHelper.getReadableDatabase();

		WhereSelectionBuilder builder = new WhereSelectionBuilder();
		if(bbox != null)
		{
			builder.appendAnd("(" + CreateNoteTable.Columns.LATITUDE + " BETWEEN ? AND ?)",
					String.valueOf(bbox.getMinLatitude()),
					String.valueOf(bbox.getMaxLatitude()));
			builder.appendAnd("(" + CreateNoteTable.Columns.LONGITUDE + " BETWEEN ? AND ?)",
					String.valueOf(bbox.getMinLongitude()),
					String.valueOf(bbox.getMaxLongitude()));
		}

		List<CreateNote> result = new ArrayList<>();

		try (Cursor cursor = db.query(CreateNoteTable.NAME, null, builder.getWhere(), builder.getArgs(),
				null, null, null, null))
		{
			if (cursor.moveToFirst())
			{
				while (!cursor.isAfterLast())
				{
					result.add(createObjectFrom(cursor));
					cursor.moveToNext();
				}
			}
		}

		return result;
	}

	private CreateNote createObjectFrom(Cursor cursor)
	{
		int colNoteId = cursor.getColumnIndexOrThrow(CreateNoteTable.Columns.ID),
			colLat = cursor.getColumnIndexOrThrow(CreateNoteTable.Columns.LATITUDE),
			colLon = cursor.getColumnIndexOrThrow(CreateNoteTable.Columns.LONGITUDE),
			colText = cursor.getColumnIndexOrThrow(CreateNoteTable.Columns.TEXT),
			colElementType = cursor.getColumnIndexOrThrow(CreateNoteTable.Columns.ELEMENT_TYPE),
			colElementId = cursor.getColumnIndexOrThrow(CreateNoteTable.Columns.ELEMENT_ID),
			colQuestTitle = cursor.getColumnIndexOrThrow(CreateNoteTable.Columns.QUEST_TITLE),
			colImagePaths = cursor.getColumnIndexOrThrow(CreateNoteTable.Columns.IMAGE_PATHS);

		CreateNote note = new CreateNote();
		note.position = new OsmLatLon(cursor.getDouble(colLat), cursor.getDouble(colLon));
		note.text = cursor.getString(colText);
		if(!cursor.isNull(colQuestTitle))
		{
			note.questTitle = cursor.getString(colQuestTitle);
		}
		if(!cursor.isNull(colElementType))
		{
			note.elementType = Element.Type.valueOf(cursor.getString(colElementType));
		}
		if(!cursor.isNull(colElementId))
		{
			note.elementId = cursor.getLong(colElementId);
		}
		if(!cursor.isNull(colImagePaths))
		{
			note.imagePaths = serializer.toObject(cursor.getBlob(colImagePaths), ArrayList.class);
		}
		note.id = cursor.getLong(colNoteId);

		return note;
	}

}

File: app/src/main/java/de/westnordost/streetcomplete/data/osmnotes/CreateNoteDao.kt
package de.westnordost.streetcomplete.data.osmnotes

import android.database.Cursor
import android.database.sqlite.SQLiteOpenHelper
import androidx.core.content.contentValuesOf

import javax.inject.Inject

import de.westnordost.streetcomplete.data.WhereSelectionBuilder
import de.westnordost.osmapi.map.data.BoundingBox
import de.westnordost.osmapi.map.data.Element
import de.westnordost.osmapi.map.data.OsmLatLon
import de.westnordost.streetcomplete.data.ObjectRelationalMapping
import de.westnordost.streetcomplete.data.osm.ElementKey
import de.westnordost.streetcomplete.data.osmnotes.CreateNoteTable.Columns.ELEMENT_ID
import de.westnordost.streetcomplete.data.osmnotes.CreateNoteTable.Columns.ELEMENT_TYPE
import de.westnordost.streetcomplete.data.osmnotes.CreateNoteTable.Columns.ID
import de.westnordost.streetcomplete.data.osmnotes.CreateNoteTable.Columns.IMAGE_PATHS
import de.westnordost.streetcomplete.data.osmnotes.CreateNoteTable.Columns.LATITUDE
import de.westnordost.streetcomplete.data.osmnotes.CreateNoteTable.Columns.LONGITUDE
import de.westnordost.streetcomplete.data.osmnotes.CreateNoteTable.Columns.QUEST_TITLE
import de.westnordost.streetcomplete.data.osmnotes.CreateNoteTable.Columns.TEXT
import de.westnordost.streetcomplete.data.osmnotes.CreateNoteTable.NAME
import de.westnordost.streetcomplete.ktx.*
import de.westnordost.streetcomplete.util.Serializer

class CreateNoteDao @Inject constructor(
    private val dbHelper: SQLiteOpenHelper,
    private val mapping: CreateNoteMapping
) {
    private val db get() = dbHelper.writableDatabase

    fun add(note: CreateNote): Boolean {
        val rowId = db.insert(NAME, null, mapping.toContentValues(note))
        if (rowId == -1L) return false

        note.id = rowId
        return true
    }

    fun get(id: Long): CreateNote? {
        return db.queryOne(NAME, null, "$ID = $id") { mapping.toObject(it) }
    }

    fun getCount(): Int {
        return db.queryOne(NAME, arrayOf("COUNT(*)")) { it.getInt(0) } ?: 0
    }

    fun delete(id: Long): Boolean {
        return db.delete(NAME, "$ID = $id", null) == 1
    }

    fun getAll(): List<CreateNote> {
        return db.query(NAME) { mapping.toObject(it) }
    }

    fun getAll(bbox: BoundingBox): List<CreateNote> {
        val builder = WhereSelectionBuilder()
        builder.appendBounds(bbox)

        return db.query(NAME, null, builder.where, builder.args) { mapping.toObject(it) }
    }
}

private fun WhereSelectionBuilder.appendBounds(bbox: BoundingBox) {
    add("($LATITUDE BETWEEN ? AND ?)",
        bbox.minLatitude.toString(),
        bbox.maxLatitude.toString()
    )
    add(
        "($LONGITUDE BETWEEN ? AND ?)",
        bbox.minLongitude.toString(),
        bbox.maxLongitude.toString()
    )
}

class CreateNoteMapping @Inject constructor(private val serializer: Serializer)
    : ObjectRelationalMapping<CreateNote> {

    override fun toContentValues(obj: CreateNote) = contentValuesOf(
        LATITUDE to obj.position.latitude,
        LONGITUDE to obj.position.longitude,
        ELEMENT_TYPE to obj.elementKey?.elementType?.name,
        ELEMENT_ID to obj.elementKey?.elementId,
        IMAGE_PATHS to obj.imagePaths?.let { serializer.toBytes(ArrayList(it)) },
        TEXT to obj.text,
        QUEST_TITLE to obj.questTitle
    )

    override fun toObject(cursor: Cursor) = CreateNote(
        cursor.getLong(ID),
        cursor.getString(TEXT),
        OsmLatLon(cursor.getDouble(LATITUDE), cursor.getDouble(LONGITUDE)),
        cursor.getStringOrNull(QUEST_TITLE),
        cursor.getStringOrNull(ELEMENT_TYPE)?.let { type ->
            cursor.getLongOrNull(ELEMENT_ID)?.let { id ->
                ElementKey(Element.Type.valueOf(type), id)
            }
        },
        cursor.getBlobOrNull(IMAGE_PATHS)?.let { serializer.toObject<ArrayList<String>>(it) }
    )
}
